source/common/http/http2/codec_impl.cc

```diff
@@ -467,8 +467,11 @@ ConnectionImpl::ConnectionImpl(Network::Connection& connection, Stats::Scope& st
 
 ConnectionImpl::~ConnectionImpl() { nghttp2_session_del(session_); }
 
-void ConnectionImpl::dispatch(Buffer::Instance& data) {
+Http::Status ConnectionImpl::dispatch(Buffer::Instance& data) {
   ENVOY_CONN_LOG(trace, "dispatching {} bytes", connection_, data.length());
+  // Make sure that dispatching_ is set to false after dispatching, even when
+  // ConnectionImpl::dispatch throws an exception.
+  Cleanup cleanup([this]() { dispatching_ = false; });
```

Comment:
TODO to make this not throw and thus remove this hack? It seems unexpected to have a method that returns a status and can also throw.

Reply:
done

Reply:
I think you are still going to need this given early returns?

Reply:
Even with early returns, we would still be executing a `dispatching_ = false` line at the end of the method if we are guaranteed not to throw, right? Or early returns like possible `RELEASE_ASSERT`s? If so I'll add a clarification

Reply:
I just meant early returns in this function, unless you structure it in such a way to only have a single return at the end. It's not a big deal either way, we can sort it in future PRs?

Reply:
Does this function still call others that throw? Or are the throws only self-inflicted in the loop below? Maybe we can just clean this up right here?

Reply:
Gotcha -- yeah, it'll depend on the next PR. I clarified, left the issue # in there just for remembering in future PRs when it comes up!

Reply:
> Does this function still call others that throw?
Yeah -- `nghttp2_session_mem_recv` may throw in a callback, so I don't think it's possible to clean up right now.